aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    constant uint3             & sizes          [[buffer(4)]],
+    threadgroup char           * shared_memory  [[threadgroup(0)]], // threadgroup buffer at index 0
+    uint3                        tgpig          [[threadgroup_position_in_grid]], // 3d coordinates
+    uint                         tiitg          [[thread_index_in_threadgroup]], // 128 per threadgroup
+    uint                         sgitg          [[simdgroup_index_in_threadgroup]]) {
+
+    using T4 = typename BlockType<T>::type4;
+    using Tsimd8x8 = typename BlockType<T>::simdgroup_type8x8;
+    // sizes: x = M, y = K, z = N
+    // pytorch: M x K @ N x K -> M x N
+    // ggml: K x N @ K x M -> N x M
+    uint32_t M = sizes.x; // M
+    uint32_t K = sizes.y; // K
+    uint32_t N = sizes.z; // N
+    uint32_t nb00 = sizeof(W);
+    uint32_t nb01 = nb00 * K;
+    uint32_t nb10 = sizeof(T);
+    uint32_t nb11 = nb10 * K;
+
+    // 8192 for sa, 4096 for sb
+    threadgroup T      * sa = (threadgroup T      *)(shared_memory);
+    threadgroup bfloat * sb = (threadgroup bfloat *)(shared_memory + 8192);
+
+    const uint r0 = tgpig.y;
+    const uint r1 = tgpig.x;
+
+    // if this block is of 64x32 shape or smaller
+    short n_rows_A = (M - r1 * BLOCK_SIZE_M < BLOCK_SIZE_M) ? (M - r1 * BLOCK_SIZE_M) : BLOCK_SIZE_M;
+    short n_rows_B = (N - r0 * BLOCK_SIZE_N < BLOCK_SIZE_N) ? (N - r0 * BLOCK_SIZE_N) : BLOCK_SIZE_N;
+
+    // a thread shouldn't load data outside of the matrix
+    short thread_row_A = ((short)tiitg/THREAD_PER_ROW_A) < n_rows_A
+        ? ((short)tiitg/THREAD_PER_ROW_A)
+        : n_rows_A - 1;
+    short thread_row_B = ((short)tiitg/THREAD_PER_ROW_B) < n_rows_B
+        ? ((short)tiitg/THREAD_PER_ROW_B)
+        : n_rows_B - 1;
+
+    Tsimd8x8 ma[2]; // input
+    simdgroup_bfloat8x8 mb[4]; // weight
+    simdgroup_float8x8 c_res[8]; // outer product result
+    for (short i = 0; i < 8; i++){
+        c_res[i] = make_filled_simdgroup_matrix<float, 8>(0.f);
+    }
+
+    constant T * a_ptr = (constant T *)((constant char *)A
+        + nb11 * (r1 * BLOCK_SIZE_M + thread_row_A)
+        + nb10 * (BLOCK_SIZE_K / THREAD_PER_ROW_A * (tiitg % THREAD_PER_ROW_A)));
+
+    constant W * b_ptr = (constant W *)(B
+        + nb01 * (r0 * BLOCK_SIZE_N + thread_row_B)
+        + nb00 * (BLOCK_SIZE_K / THREAD_PER_ROW_B * (tiitg % THREAD_PER_ROW_B)));
+/**
+Load weight and input into shared memory:
+
+             K
+  ┌────────────────────────┐              8192 (A)           4096 (B)
+  │                        │   ┌────────────────────────┬────────────┐
+  │                        │   │++++++++++++++++++++++++│++++++++++++│
+  │                        │   └────────────────────────┴────────────┘
+  │                        │
+  │32                      │
+  ├──┬──┬──────────────────┤                    K
+  │++│  │                  │        ┌────────────────────────┐
+64│++│  │  ...             │        │                        │
```

Comment:
can `BLOCK_SIZE_M = 32`& `BLOCK_SIZE_N = 64` appear in the diagram so that it is easier to track the code and the diagram together?